### PR TITLE
Canbus mcp2515 test for integrated review

### PR DESCRIPTION
**Pull request by mvturnho**
## Description:

Joined canbus and mcp2515

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/648

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by mvturnho** [comment]
@OttoWinter In this repo you can check the compile, lint and other checks. I also tested this PR on the test hardware and all functions good.
I then copied the code from this branch into the separate mcp2515 an canbus_base branches.

but I think you have more control and assurance in commiting this PR then the separate mcp2515 and canbus_base components.

**Comment by brandond** [comment]
This looks good to me. If it works on your test hardware then I'd say lets merge it.

**Review by brandond — APPROVED** [review]

**Comment by mvturnho** [comment]
> LGTM.
> 
> If it works on your test hardware then I'd say lets merge it.

@brandond  This slipt my attantion, thank you I will add the documentation as well. I got this integration running for several months and it seems real stable.

File: esphome/components/canbus/__init__.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+CONF_ON_FRAME = 'on_frame'
+
+CONF_CANBUS_SEND = 'canbus.send'
+
+
+def validate_raw_data(value):
+    if isinstance(value, str):
+        return value.encode('utf-8')
+    if isinstance(value, str):
+        return value
```
Comment (glmnet) [inline esphome/components/canbus/__init__.py:22]: not a python expert, but this code is intended as is? you are checking twice the same here

**Comment by glmnet** [comment]
@mvturnho a bit off, but have you seen this already https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/can.html

in that case do you think base can will accept an implementation for this for esp32?

**Comment by glmnet** [comment]
How are you wiring the MCP2515 to the ESP8266?
Any level converter?

**Comment by danielschramm** [comment]
> How are you wiring the MCP2515 to the ESP8266?
> Any level converter?

Hello,

first option is to get an expensive 3.3V Board which runs the MCP with 3.3V and uses a special CAN Transceiver.

But usually connecting a 5V SPI Device to any 3.3V Microcontroller is not a bis issue if supply tolerances are within small ranges (usually the case in most applications)
The 3.3V Output of the Microcontroller can be directly connected to the Input at a 5V Device.
The other way round the 5V Level could damage the Microcontroller so the inrush current has to be limited by a series resistor. If the device produces a good low level no divider is needed only a series resistor.
Internal Diodes inside the Microcontroller will then limit the Signal to 3.3V. 10kOhm is a good rule of thumb depending on the signal speed.

I will do some testing and schematics once this request is integrated into the main tree. And may provide some wiring diagrams on success to integrate in the docs.

In rare cases this fails and some more effort is needed. 
Reasons for that are well explained on this site:
https://next-hack.com/index.php/2020/02/15/how-to-interface-a-3-3v-output-to-a-5v-input/

**Comment by glmnet** [comment]
Well.. Thanks for the electronics background

Indeed I just connected it, and feeded the MCP2515 with 5V directly, it's working on the ESP8266, I believe the ESP32 can handle it equally right, did not test it though.

I have some level converters but they seems to be two channels, but tx and rx separated, so it's not two way. I didn't check how SPI really works so I just went the straight way.

On the Raspberry PI they say it cannot tolerate 5V on the SPI pins, so I hacked a MCP2515 board, the MCP2515 works ok at 3.3V and can connect to the CAN transceivers (TJA1050) even when the transceivers is powered via 5V so I feeded 5V only to the transceiver, that is working too. (I could have done the same for the ESP but doesn't *seem* necessary)

BTW, the ESP32 has a CAN bus controller built in, so only a CAN bus transceiver is needed, I am using it without problems too but SW is not an ESPHome component. There is a 3.3V transceiver for CAN bus which can be directly connected to the ESP32. (SN65HVD)

**Comment by danielschramm** [comment]
Esp32 and esp8266 are not officially 5V input tolerant.
There are a lot of discussions about that.
Some have it working with 5V input other burnt the chips.
It seems that the input just can withstand relatively high input clamping current and it depends on the driver strength if the chip survives. Also different Pins can have different input circuit if they provide different functionality.
If the communication is running well without any resistor it might not be guaranteed for a long lifetime but then the Solution with just series resistors will work perfectly as you confirmed the voltage levels match. The resistor just needs to limit the current to prevent damage.

**Comment by glmnet** [comment]
Ok, so I better use the 10K Ohm there, should be on all the MOSI, MISO, SCK and CS pins right? 
What do you think about running the MCP2515 at 3.3V and the transceiver at 5V?

**Comment by danielschramm** [comment]
About 3.3V transceiver to the ESP32.
On the first hand it sounds interesting but these full 3.3V Chips have a lower BUS voltage and cannot fulfil the specified cable length CAN normally provides.
There are some better drivers out there with split supply 3.3V logic and 5V CAN side like the AMIS-30663. That is the best solution in professional use but hard to source for hobby.


Hint for decision in home use:
- 5V available: use a cheap 5V Transceiver. Connect TX direct und RX with series Resistor. Gives better CAN Levels and lowest cost.
- Only 3.3V available: use a 3.3V Transceiver but reduce cable length or bitrate.

**Comment by danielschramm** [comment]
Only a resistor on MISO is enough since this is the only signal driven with 5V logic level.
MOSI, SCK and CS are driven by the ESP with 3.3V.

The answer to running transceiver with other voltage than the CAN controller is just answered in my last post.
In this special case it could be done but does not provide a real advantage. Only the current consumption of the MSP would be little lower since the current is linear to voltage on cmos logic.

**Comment by glmnet** [comment]
@mvturnho I forced pushed this one as I had to sync it with dev due to a change in SPI. Also did some work on docs and config schema. Do you want to check out?

It's GTM for me

**Comment by mvturnho** [comment]
@glmnet hi,thanks for this. I will checkout the docs and comment if needed.

**Comment by danielschramm** [comment]
Hello mvturnho, hello glmnet,

could you please provide the pinout which you are successfully running on your boards?
I would like to do some Schematic drawings which could be added to the docs.
I know SPI allows different combination, but for best practice I like to document a tested combination.
Did you also use the INT pin of the MCP2515?

Thanks, Daniel

**Comment by glmnet** [comment]
```yaml

spi:
  id: spi1
  clk_pin: D0
  mosi_pin: D1
  miso_pin: D2


canbus:
  - platform: mcp2515
    cs_pin: D5
    can_id: 4
    bit_rate: 50kbps
```

thats my test so far, nothing connected to `INT`

**Comment by glmnet** [comment]
I have a 10K resistor in series on the MISO line, it works, here is a scope capture, yellow is MCP2515 side and blue is ESP8266 side

pic_54_8

**Comment by glmnet** [comment]
Now I added a 22k pull down to configure a voltage divider and got this, all other lines are in the 3.3V range, looks like the MCP2515 is happy, at least at 50Kbps, we can document this as a safe configuration, what do you think @danielschramm 

pic_54_9

**Comment by danielschramm** [comment]
The first shot shows that the Pin at least tries to be 5V  compliant since it does not clamp the signal to 3.3V as a normal CMOS input would do.
I will do some testing on that when back from holiday.
(Currently I have only very limited equipment in my laptop bag)

The voltage divider option is implemented with really huge values which is OK for 50kBit/s but might fail at 1Mbit/s. I need to check the input capacitance to calculate a high speed capable divider.

I will calculate and document some options and we then can collect user feedback and keep successful implementations.

**Comment by danielschramm** [comment]
Hello,

I attached some drafts for schematic documentation.
The pictures in in schematics are not checked for proper license I ordered all the modules to test and take own photos afterwards it could be integrated into the official documentation.

If someone likes to get the current draft as openoffice or kicad draft, please send a personal message.

[esphome_3v3_draft.pdf](https://github.com/esphome/esphome/files/4908813/esphome_3v3_draft.pdf)
[can_interfaces_draft.pdf](https://github.com/esphome/esphome/files/4908814/can_interfaces_draft.pdf)

**Comment by loongyh** [comment]
Regarding the 3.3V vs 5V issue, this [TI report](https://www.ti.com/lit/an/slla337/slla337.pdf) says they're inconsequential and interoperable as both 3.3V and 5V drivers provide the differential voltage of >=1.5V between CANL and CANH.

**Comment by danielschramm** [comment]
Hello,

does someone know the status of this pull request.
The referenced branch to the pull seems to be deleted and it seems to be not merged yet?
Is it worth updating the docs I started in the summer?
If it is not yet stable enough for a dev merge could you start an official branch to harden this integration together?

I usually have some time during winter to do at least some testing and doc work.

Thanks,

Daniel

**Comment by danielschramm** [comment]
Hello,

the "needs-doc" flag could be removed.
The documentation from Michiel was already really good. I just added 2 wiring diagrams to remove the last TODO in the Pull request to his docs:
https://github.com/mvturnho/esphome-docs/pull/13

What work needs to be done to get this merged?

Thanks,

Daniel

**Comment by danielschramm** [comment]
Hello,

I just tested if the code from this request is still working with the current esphome/dev.
Everything seems ok, it compiles and is fully functional.
https://github.com/danielschramm/esphome/tree/canbus

Daniel

**Comment by danielschramm** [comment]
Hello,

I did a lot of testing on this feature and found some minor issues in the code and would like to invest some more time to get this great work from mvturnho merged finally.
So I would like to suggest closing this pull request and the corresponding docs https://github.com/esphome/esphome-docs/pull/648 and create  new ones based on my fork to which I already invited mvturnho.

Thanks,
Daniel

**Comment by glmnet** [comment]
looks like @mvturnho branch was removed or something, @danielschramm are you creating a proper PR? can you link them here?

**Comment by danielschramm** [comment]
> looks like @mvturnho branch was removed or something, @danielschramm are you creating a proper PR? can you link them here?

Ok, just created: https://github.com/esphome/esphome/pull/1337 and https://github.com/esphome/esphome-docs/pull/823

**Comment by mvturnho** [comment]
> > looks like @mvturnho branch was removed or something, @danielschramm are you creating a proper PR? can you link them here?
> 
> Ok, just created: #1337 and [esphome/esphome-docs#823](https://github.com/esphome/esphome-docs/pull/823)

Hi I see there is activity on de canbus integration I made months ago. Please tell me what you need to gave this committed. I have this running for several months now and the code seems stable.

**Comment by danielschramm** [comment]
> > > looks like @mvturnho branch was removed or something, @danielschramm are you creating a proper PR? can you link them here?
> > 
> > 
> > Ok, just created: #1337 and [esphome/esphome-docs#823](https://github.com/esphome/esphome-docs/pull/823)
> 
> Hi I see there is activity on de canbus integration I made months ago. Please tell me what you need to have this committed. I have this running for several months now and the code seems stable.

Hello,

the payload was not correctly passed to the lambda only length-1. I fixed it in the new pull.
And I replaced your left TODO in the Docs.
Apart from these really small things I'm just about getting my window shades ported to work together with your integration.
(http://pinguin.sauerland.de/din_rail.html)

We should close this Pull here since the new one passed all the tests. I invited you to my forks that we can work together on everything the reviewers will find.

Thanks a lot for your work.

Daniel

**Comment by glmnet** [comment]
> > > looks like @mvturnho branch was removed or something, @danielschramm are you creating a proper PR? can you link them here?
> > 
> > 
> > Ok, just created: #1337 and [esphome/esphome-docs#823](https://github.com/esphome/esphome-docs/pull/823)
> 
> Hi I see there is activity on de canbus integration I made months ago. Please tell me what you need to have this committed. I have this running for several months now and the code seems stable.

Looks like your branch for this PR is deleted or something? I see ``unknown repository`` on top, and if I click this url gaves me 404 https://github.com/esphome/esphome/tree/canbus_mcp2515_TEST

This is why I decided to create a new PR, you should be credited though.

**Comment by danielschramm** [comment]
> > > > looks like @mvturnho branch was removed or something, @danielschramm are you creating a proper PR? can you link them here?
> > > 
> > > 
> > > Ok, just created: #1337 and [esphome/esphome-docs#823](https://github.com/esphome/esphome-docs/pull/823)
> > 
> > 
> > Hi I see there is activity on de canbus integration I made months ago. Please tell me what you need to have this committed. I have this running for several months now and the code seems stable.
> 
> Looks like your branch for this PR is deleted or something? I see `unknown repository` on top, and if I click this url gaves me 404 https://github.com/esphome/esphome/tree/canbus_mcp2515_TEST
> 
> This is why I decided to create a new PR, you should be credited though.

I updated the comment of the new request to make clear who really created the code!

**Pull request closed without merging**